# 4.8 Exercise - Grouping Data and Aggregating Variables 

Table of Contents

01. Importing

02. Grouping

03. Aggregating with agg()

04. Aggregating with transform()

05. Task: Step 2-7
    
06. Export Data

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

Importing data frame

In [2]:
path = r'C:\Users\31622\OneDrive\Documents\Career Foundry\Instacart Basket Analysis'

In [3]:
# Importing dataframe 
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_updated.pkl'))

In [4]:
# Check shape of dataframe 
ords_prods_merge.shape

(32434212, 18)

In [5]:
# Check dataframe structure
ords_prods_merge.head()

order_id  user_id  number_of_orders  orders_day_of_week  order_hour_of_day  \
0   2539329        1                 1                   2                  8   
1   2398795        1                 2                   3                  7   
2    473747        1                 3                   3                 12   
3   2254736        1                 4                   4                  7   
4    431534        1                 5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      busiest_day  busiest_days busiest_period_of_day  
0  Regularly busy  Regular days        Average orders  
1  Regularly busy  Slowest days        Average orders  
2  Regularly busy  Slowest days           Most orders  
3      Least busy  Slowest days        Average orders  
4      Least busy  Slowest days           Most orders

In [6]:
# check columns in dataframe
ords_prods_merge.columns

Index(['order_id', 'user_id', 'number_of_orders', 'orders_day_of_week',
       'order_hour_of_day', 'days_since_prior_order', 'product_id',
       'add_to_cart_order', 'reordered', 'product_name', 'aisle_id',
       'department_id', 'prices', '_merge', 'price_range_loc', 'busiest_day',
       'busiest_days', 'busiest_period_of_day'],
      dtype='object')

Question 2: Find the mean orders for each department in the merged dataframe

In [7]:
# Group by "product_name" 
ords_prods_merge.groupby('product_name')

In [8]:
#Split the data into groups based on “department_id.”
#Apply the agg() function to each group to obtain the mean values for the “number_of_orders” column.
ords_prods_merge.groupby('department_id').agg({'number_of_orders': ['mean']})

number_of_orders
                          mean
department_id                 
1                    15.457687
2                    17.277920
3                    17.179756
4                    17.811403
5                    15.213779
6                    16.439806
7                    17.225773
8                    15.340520
9                    15.895474
10                   20.197148
11                   16.170828
12                   15.887622
13                   16.583304
14                   16.757377
15                   16.165037
16                   17.663250
17                   15.694469
18                   19.310397
19                   17.177343
20                   16.473447
21                   22.902379

In [12]:
# Importing department_wrangled dataframe
df_dept = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'departments_wrangled.csv'))
# Importing dataframe to help better understand the different departments in our analysis. 

In [13]:
df_dept

Unnamed: 0       department
0            1           frozen
1            2            other
2            3           bakery
3            4          produce
4            5          alcohol
5            6    international
6            7        beverages
7            8             pets
8            9  dry goods pasta
9           10             bulk
10          11    personal care
11          12     meat seafood
12          13           pantry
13          14        breakfast
14          15     canned goods
15          16       dairy eggs
16          17        household
17          18           babies
18          19           snacks
19          20             deli
20          21          missing

Answer 2: Compared to our subset from our practise exercise, we are able to see each department and the mean for this.
Now we have our full dataframe, we see that the mean recorded for our subset has changed, with less variablity perhaps a better reflecton of variablity of the mean orders.
for our dataframe. We see in our entire dataframe that the department with minimum mean orders is alcohol (15.2) and the maximum mean orders is missing categories (22.9).
This missing category indicates a large number of unspecified values for the column "department_id."

Question 4: Create a loyalty flag for existing customers using the transform and loc functions.  

In [14]:
# groups based by the “user_id” column, generate maximum orders for each user, create “max_order,” results of aggregation.
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['number_of_orders'].transform(np.max)

In [15]:
#Check dataframe
ords_prods_merge.head()

order_id  user_id  number_of_orders  orders_day_of_week  order_hour_of_day  \
0   2539329        1                 1                   2                  8   
1   2398795        1                 2                   3                  7   
2    473747        1                 3                   3                 12   
3   2254736        1                 4                   4                  7   
4    431534        1                 5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      busiest_day  busiest_days busiest_period_of_day  max_order  
0  Regularly busy  Regular days        Average orders         10  
1  Regularly busy  Slowest days        Average orders         10  
2  Regularly busy  Slowest days           Most orders         10  
3      Least busy  Slowest days        Average orders         10  
4      Least busy  Slowest days           Most orders         10

In [16]:
# Check the data grouping for accuracy
pd.options.display.max_rows = None

In [17]:
ords_prods_merge.head(100)

order_id  user_id  number_of_orders  orders_day_of_week  \
0    2539329        1                 1                   2   
1    2398795        1                 2                   3   
2     473747        1                 3                   3   
3    2254736        1                 4                   4   
4     431534        1                 5                   4   
5    3367565        1                 6                   2   
6     550135        1                 7                   1   
7    3108588        1                 8                   1   
8    2295261        1                 9                   1   
9    2550362        1                10                   4   
10   2968173       15                15                   1   
11   1870022       15                17                   2   
12   1911383       15                18                   2   
13   2715276       15                21                   1   
14    487368       15                22                   1   
15   2293453       19                 2                   5   
16   1973799       19                 5                   6   
17    532817       19                 7                   4   
18   1573906       21                10                   3   
19   1593000       31                10                   3   
20   2231262       31                17                   3   
21   2580647       43                 6                   4   
22   2187180       43                 9                   4   
23   2497897       52                 1                   1   
24   1318871       52                 2                   1   
25   1261384       52                 3                   1   
26    580568       52                 5                   2   
27   2428073       52                 7                   2   
28    342306       52                 9                   1   
29   1520353       52                10                   2   
30   2542086       52                11                   2   
31    944694       52                12                   1   
32   1268191       52                15                   2   
33   1122089       52                17                   2   
34   1498922       52                18                   2   
35    180919       52                19                   3   
36   2695875       52                21                   5   
37   3205960       67                 1                   3   
38    970922       67                 2                   1   
39    596585       67                 3                   1   
40   2037087       67                 4                   2   
41   2474111       67                 5                   3   
42   1568481       67                 6                   3   
43   2048527       67                 8                   2   
44   2389031       67                 9                   2   
45   2378562       67                10                   1   
46   1515698       67                11                   1   
47   1928050       67                12                   2   
48   1949274       67                13                   1   
49    551869       67                14                   1   
50    768572       67                16                   4   
51    150528       67                18                   5   
52     29758       67                19                   0   
53   1558976       67                21                   1   
54    863974       67                22                   1   
55   2308399       67                23                   1   
56   2759128       81                 3                   1   
57   1812985       81                 5                   2   
58   1889835       82                 1                   3   
59    499460       82                 3                   5   
60   2728456       82                 4                   0   
61   2653743       82                 6                   4   
62   1358880       82              

Question 5: Create flags for loyalty labels 

In [18]:
# Create flags to assign "loyalty" labels to user ID based on max order value
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [19]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [20]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [21]:
# Check frequencies of loyalty flags for accuracy
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

Regular customer    15891077
Loyal customer      10293737
New customer         6249398
Name: loyalty_flag, dtype: int64

In [22]:
ords_prods_merge[['user_id', 'loyalty_flag', 'number_of_orders']].head(20)

user_id      loyalty_flag  number_of_orders
0         1      New customer                 1
1         1      New customer                 2
2         1      New customer                 3
3         1      New customer                 4
4         1      New customer                 5
5         1      New customer                 6
6         1      New customer                 7
7         1      New customer                 8
8         1      New customer                 9
9         1      New customer                10
10       15  Regular customer                15
11       15  Regular customer                17
12       15  Regular customer                18
13       15  Regular customer                21
14       15  Regular customer                22
15       19      New customer                 2
16       19      New customer                 5
17       19      New customer                 7
18       21  Regular customer                10
19       31  Regular customer                10

Question 5: Compare spending habits across the three types of customers (loyalty groups)

In [23]:
# Group by "prices" 
ords_prods_merge.groupby('prices')

In [24]:
# Group data based on “loyalty_flag.”
# Apply the agg() function to obtain the mean values for the “prices” column
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean']})

prices
                       mean
loyalty_flag               
Loyal customer    10.388747
New customer      13.294370
Regular customer  12.496203

Answer 5: As shown here, the mean prices of products purchased by loyal customers falls below that of new customers and regular customer.

Question 6: Create a spending flag for each user based on the average price across all their orders - Categorise as 'Low spender' and 'High spender' based on the crieria given

In [25]:
# Group data by "user_id" column, generate mean product price for each user, create "mean_price" for aggregation results
ords_prods_merge['mean_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.max)

In [26]:
# Create flags to assign "Spending" labels based on mean prices value 
ords_prods_merge.loc[ords_prods_merge['mean_price'] < 10, 'Spending_flag'] = 'Low spender'

In [27]:
ords_prods_merge.loc[ords_prods_merge['mean_price'] >= 10, 'Spending_flag'] = 'High spender'

In [28]:
# Check frequencies of spending flags for accuracy
ords_prods_merge['Spending_flag'].value_counts(dropna = False)

High spender    32402635
Low spender        31577
Name: Spending_flag, dtype: int64

In [29]:
ords_prods_merge.head(20)

order_id  user_id  number_of_orders  orders_day_of_week  \
0    2539329        1                 1                   2   
1    2398795        1                 2                   3   
2     473747        1                 3                   3   
3    2254736        1                 4                   4   
4     431534        1                 5                   4   
5    3367565        1                 6                   2   
6     550135        1                 7                   1   
7    3108588        1                 8                   1   
8    2295261        1                 9                   1   
9    2550362        1                10                   4   
10   2968173       15                15                   1   
11   1870022       15                17                   2   
12   1911383       15                18                   2   
13   2715276       15                21                   1   
14    487368       15                22                   1   
15   2293453       19                 2                   5   
16   1973799       19                 5                   6   
17    532817       19                 7                   4   
18   1573906       21                10                   3   
19   1593000       31                10                   3   

    order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                   8                     NaN         196                  1   
1                   7                    15.0         196                  1   
2                  12                    21.0         196                  1   
3                   7                    29.0         196                  1   
4                  15                    28.0         196                  1   
5                   7                    19.0         196                  1   
6                   9                    20.0         196                  1   
7                  14                    14.0         196                  2   
8                  16                     0.0         196                  4   
9                   8                    30.0         196                  1   
10                  9                     7.0         196                  2   
11                 16                     8.0         196                  6   
12                 11                     7.0         196                  1   
13                  9                     7.0         196                  1   
14                 10                    14.0         196                  1   
15                 14                     6.0         196                  3   
16                 12                     8.0         196                 15   
17                 17                     6.0         196                  1   
18                 10                     6.0         196                  2   
19                  8                     7.0         196                 17   

    reordered product_name  ...  prices  _merge    price_range_loc  \
0           0         Soda  ...     9.0    both  Mid-range product   
1           1         Soda  ...     9.0    both  Mid-range product   
2           1         Soda  ...     9.0    both  Mid-range product   
3           1         Soda  ...     9.0    both  Mid-range product   
4           1         Soda  ...     9.0    both  Mid-range product   
5           1         Soda  ...     9.0    both  Mid-range product   
6           1         Soda  ...     9.0    both  Mid-range product   
7           1         Soda  ...     9.0    both  Mid-range product   
8           1         Soda  ...     9.0    both  Mid-range product   
9           1         Soda  ...     9.0    both  Mid-range product   
10          0         Soda  ...     9.0    both  Mid-range product   
11          1         Soda  ...     9.0    both  Mid-range product   
12          1         Soda  ...     9.0    both  Mid-range product   
13          1      

In [32]:
# Display the dataframe after dropping duplicated column
ords_prods_merge.columns

Index(['order_id', 'user_id', 'number_of_orders', 'orders_day_of_week',
       'order_hour_of_day', 'days_since_prior_order', 'product_id',
       'add_to_cart_order', 'reordered', 'product_name', 'aisle_id',
       'department_id', 'prices', '_merge', 'price_range_loc', 'busiest_day',
       'busiest_days', 'busiest_period_of_day', 'max_order', 'loyalty_flag',
       'mean_price', 'Spending_flag'],
      dtype='object')

Questions 7: Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column.

If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”
If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”
If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

In [33]:
# Group data by "user_id" column, generate median days for each user, create "median_days" for aggregation results
ords_prods_merge['median_days'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.max)

In [34]:
# Check grouping
ords_prods_merge.head(100)

order_id  user_id  number_of_orders  orders_day_of_week  \
0    2539329        1                 1                   2   
1    2398795        1                 2                   3   
2     473747        1                 3                   3   
3    2254736        1                 4                   4   
4     431534        1                 5                   4   
5    3367565        1                 6                   2   
6     550135        1                 7                   1   
7    3108588        1                 8                   1   
8    2295261        1                 9                   1   
9    2550362        1                10                   4   
10   2968173       15                15                   1   
11   1870022       15                17                   2   
12   1911383       15                18                   2   
13   2715276       15                21                   1   
14    487368       15                22                   1   
15   2293453       19                 2                   5   
16   1973799       19                 5                   6   
17    532817       19                 7                   4   
18   1573906       21                10                   3   
19   1593000       31                10                   3   
20   2231262       31                17                   3   
21   2580647       43                 6                   4   
22   2187180       43                 9                   4   
23   2497897       52                 1                   1   
24   1318871       52                 2                   1   
25   1261384       52                 3                   1   
26    580568       52                 5                   2   
27   2428073       52                 7                   2   
28    342306       52                 9                   1   
29   1520353       52                10                   2   
30   2542086       52                11                   2   
31    944694       52                12                   1   
32   1268191       52                15                   2   
33   1122089       52                17                   2   
34   1498922       52                18                   2   
35    180919       52                19                   3   
36   2695875       52                21                   5   
37   3205960       67                 1                   3   
38    970922       67                 2                   1   
39    596585       67                 3                   1   
40   2037087       67                 4                   2   
41   2474111       67                 5                   3   
42   1568481       67                 6                   3   
43   2048527       67                 8                   2   
44   2389031       67                 9                   2   
45   2378562       67                10                   1   
46   1515698       67                11                   1   
47   1928050       67                12                   2   
48   1949274       67                13                   1   
49    551869       67                14                   1   
50    768572       67                16                   4   
51    150528       67                18                   5   
52     29758       67                19                   0   
53   1558976       67                21                   1   
54    863974       67                22                   1   
55   2308399       67                23                   1   
56   2759128       81                 3                   1   
57   1812985       81                 5                   2   
58   1889835       82                 1                   3   
59    499460       82                 3                   5   
60   2728456       82                 4                   0   
61   2653743       82                 6                   4   
62   1358880       82              

In [35]:
# Create flags to assign "freqency" labels to user ID based on median days since prior order
ords_prods_merge.loc[ords_prods_merge['median_days'] > 20, 'frequency_flag'] = 'Non-frequent customer'

In [36]:
ords_prods_merge.loc[(ords_prods_merge['median_days'] > 10) & (ords_prods_merge['median_days'] <= 20), 'frequency_flag'] = 'Regular customer'

In [37]:
ords_prods_merge.loc[ords_prods_merge['median_days'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [38]:
# Check frequencies of spending flags for accuracy
ords_prods_merge['frequency_flag'].value_counts(dropna=False)

Non-frequent customer    24509620
Regular customer          6539471
Frequent customer         1385121
Name: frequency_flag, dtype: int64

In [39]:
ords_prods_merge.head(100)

order_id  user_id  number_of_orders  orders_day_of_week  \
0    2539329        1                 1                   2   
1    2398795        1                 2                   3   
2     473747        1                 3                   3   
3    2254736        1                 4                   4   
4     431534        1                 5                   4   
5    3367565        1                 6                   2   
6     550135        1                 7                   1   
7    3108588        1                 8                   1   
8    2295261        1                 9                   1   
9    2550362        1                10                   4   
10   2968173       15                15                   1   
11   1870022       15                17                   2   
12   1911383       15                18                   2   
13   2715276       15                21                   1   
14    487368       15                22                   1   
15   2293453       19                 2                   5   
16   1973799       19                 5                   6   
17    532817       19                 7                   4   
18   1573906       21                10                   3   
19   1593000       31                10                   3   
20   2231262       31                17                   3   
21   2580647       43                 6                   4   
22   2187180       43                 9                   4   
23   2497897       52                 1                   1   
24   1318871       52                 2                   1   
25   1261384       52                 3                   1   
26    580568       52                 5                   2   
27   2428073       52                 7                   2   
28    342306       52                 9                   1   
29   1520353       52                10                   2   
30   2542086       52                11                   2   
31    944694       52                12                   1   
32   1268191       52                15                   2   
33   1122089       52                17                   2   
34   1498922       52                18                   2   
35    180919       52                19                   3   
36   2695875       52                21                   5   
37   3205960       67                 1                   3   
38    970922       67                 2                   1   
39    596585       67                 3                   1   
40   2037087       67                 4                   2   
41   2474111       67                 5                   3   
42   1568481       67                 6                   3   
43   2048527       67                 8                   2   
44   2389031       67                 9                   2   
45   2378562       67                10                   1   
46   1515698       67                11                   1   
47   1928050       67                12                   2   
48   1949274       67                13                   1   
49    551869       67                14                   1   
50    768572       67                16                   4   
51    150528       67                18                   5   
52     29758       67                19                   0   
53   1558976       67                21                   1   
54    863974       67                22                   1   
55   2308399       67                23                   1   
56   2759128       81                 3                   1   
57   1812985       81                 5                   2   
58   1889835       82                 1                   3   
59    499460       82                 3                   5   
60   2728456       82                 4                   0   
61   2653743       82                 6                   4   
62   1358880       82              

In [40]:
# Export data to pkl
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_aggregated.pkl'))